homeassistant/components/lightwave/climate.py

```diff
@@ -0,0 +1,197 @@
+
+    for device_id, device_config in discovery_info.items():
+        name = device_config[CONF_NAME]
+        serial = device_config[CONF_SERIAL]
+        trv.append(
+            LightwaveTrv(name, device_id, lwlink, serial, trv_proxy_ip, trv_proxy_port)
+        )
+
+    async_add_entities(trv)
+
+
+class LightwaveTrv(ClimateDevice):
+    """Representation of a LightWaveRF TRV."""
+
+    def __init__(self, name, device_id, lwlink, serial, trv_proxy_ip, trv_proxy_port):
+        """Initialize LightwaveTrv entity."""
+        self._name = name
+        self._device_id = device_id
+        self._state = None
+        self._max_temp = DEFAULT_MAX_TEMP
+        self._min_temp = DEFAULT_MIN_TEMP
+        self._current_temperature = None
+        self._target_temperature = None
+        self._temperature_unit = TEMP_CELSIUS
+        self._target_temperature_step = 0.5
+        self._hvac_mode = HVAC_MODE_HEAT
+        self._hvac_action = None
+        self._lwlink = lwlink
+        self._battery = None
+        self._serial = serial
+        self._proxy_ip = trv_proxy_ip
+        self._proxy_port = trv_proxy_port
+        self._inhibit = 0
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_TARGET_TEMPERATURE
+
+    @property
+    def should_poll(self):
+        """Poll the Proxy."""
+        return True
+
+    def update_ctarg(self, j):
+        """Update target temp logic."""
+        if self._inhibit == 0:
+            self._target_temperature = j["cTarg"]
+            if j["cTarg"] == 0:
+                # TRV off
+                self._target_temperature = None
+            if j["cTarg"] >= 40:
+                # Call for heat mode, or TRV in a fixed position
+                self._target_temperature = None
+        else:
+            # Done the job - use proxy next iteration
+            self._inhibit = 0
+
+    def update(self):
+        """Communicate with a Lightwave RTF Proxy to get state."""
+        try:
+            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
+                sock.settimeout(2.0)
+                msg = self._serial.encode("UTF-8")
+                sock.sendto(msg, (self._proxy_ip, self._proxy_port))
+                response, dummy = sock.recvfrom(1024)
+                msg = response.decode()
+                j = json.loads(msg)
+                if "cTemp" in j.keys():
+                    self._current_temperature = j["cTemp"]
+                if "cTarg" in j.keys():
+                    self.update_ctarg(j)
+                if "batt" in j.keys():
+                    # convert the voltage to a rough percentage
+                    self._battery = int((j["batt"] - 2.22) * 110)
+                if "output" in j.keys():
+                    if int(j["output"]) > 0:
+                        self._hvac_action = CURRENT_HVAC_HEAT
+                    else:
+                        self._hvac_action = CURRENT_HVAC_OFF
```

Comment:
[Development checklist](https://developers.home-assistant.io/docs/en/development_checklist.html):
> All communication to external devices or services must be wrapped in an external Python library hosted on pypi.

Therefore, this socket communication needs to be moved to the library.

Reply:
My problem is I have no control over the Lightwave pypi library - I am struggling to get a pull made there as it is.

Reply:
Please open a PR for the lightwave library to include these changes. Here is a [guide from Github](https://help.github.com/en/github/getting-started-with-github/fork-a-repo) on how to fork a repo, clone and add a feature to it.

Reply:
Sorry, I'm beyond my knowledge here,
Are you referring to the lightware PyPi library, or the new lightwave proxy library?
Assuming the PyPi library...
1) I can fork the repro - and have done that, but how does that get the PyPi library updated?  I don't own it!
2) Is it just the "socket" code to get the JSON that you would like moving.   Or, is it also the logic that parses the JSON to set the various parameters.   (That would seem to be pushing HomeAssistant specific logic into the PyPi library)
Sorry for what might be basic questions, but I'm at the limit of newly found HA and github skills to get this far!

Reply:
No worries at all, feel free to ask any question 🙂
I am referring to the lightware PyPi library which is used in this component. 

> I can fork the repro - and have done that, but how does that get the PyPi library updated? I don't own it!

You are completely right just by forking and then merging a PR does not update the version of the PyPi library. This has to be updated by owner of it, therefore it is usually common courtesy to ask the owner after or during the merge to update the Pypi library as well to include these changes.

> Is it just the "socket" code to get the JSON that you would like moving. Or, is it also the logic that parses the JSON to set the various parameters. (That would seem to be pushing HomeAssistant specific logic into the PyPi library)
Sorry for what might be basic questions, but I'm at the limit of newly found HA and github skills to get this far!

Just the socket communication would have to be moved, the parsing logic can stay in HA because it is HA specific.

Reply:
Thanks.   I've moved the code to the library, and am not dependent on [GeoffAtHome](https://github.com/GeoffAtHome) accepting the pull, and updating the PyPi library.

Reply:
I think we have a misunderstanding here 😅 this entire integration is dependent on https://github.com/GeoffAtHome/lightwave and you will need to make a PR in its repo and ask for an update of the PyPi library. Otherwise, it only works on your local system but other users cannot download the altered version of `lightwave`.

Reply:
I understand, and I have made pull request on https://github.com/GeoffAtHome/lightwave, and asked for the PyPi to be updated:
https://github.com/GeoffAtHome/lightwave/pulls
As soon as that is done, we should be good to go here.

Reply:
Okay wonderful, thanks 👍 the links or mentions for this PR were missing so I was a bit confused.